packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -0,0 +1,250 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const error = require('../error');
+const subhub = require('../subhub/client');
+
+const stripe = require('stripe');
+
+/** @typedef {import('stripe').customers.ICustomer} Customer */
+/** @typedef {import('stripe').products.IProduct} Product */
+/** @typedef {import('stripe').plans.IPlan} Plan */
+/** @typedef {import('stripe').subscriptions.ISubscription} Subscription */
+/** @typedef {import('stripe').customers.ICustomerSubscriptions} Subscriptions */
+
+/**
+ * @typedef AbbrevProduct
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ */
+
+/**
+ * @typedef AbbrevPlan
+ * @property {string} plan_id
+ * @property {string} plan_name
+ * @property {Product['metadata']} plan_metadata
+ * @property {string} product_id
+ * @property {string} product_name
+ * @property {Product['metadata']} product_metadata
+ * @property {Plan['interval']} interval
+ * @property {Plan['amount']} amount
+ * @property {Plan['currency']} currency
+ */
+
+/**
+ * Get a cached result at a cache key and regenerated it with `refreshFunction`
+ * if its expired.
+ *
+ * @template T
+ * @param {*} redis
+ * @param {string} cacheKey
+ * @param {number} cacheTtl
+ * @param {() => Promise<T>} refreshFunction
+ * @returns {Promise<T>} possibly cached result
+ */
+async function cachedResult(log, redis, cacheKey, cacheTtl, refreshFunction) {
+  if (cacheTtl) {
+    try {
+      const json = await redis.get(cacheKey);
+      if (json) {
+        return JSON.parse(json);
+      }
+    } catch (err) {
+      log.error(
+        `subhub.cachedResult.${refreshFunction.name}.getCachedResponse.failed`,
+        { err }
+      );
+    }
+  }
+
+  const result = await refreshFunction();
+  if (cacheTtl) {
+    redis
+      .set(cacheKey, JSON.stringify(result), 'EX', cacheTtl)
+      .catch(err =>
+        log.error(
+          `subhub.cachedResult.${refreshFunction.name}.cacheResponse.failed`,
+          { err }
+        )
+      );
+  }
+  return result;
+}
+
+class StripeHelper {
+  /**
+   * Create a Stripe Helper with built-in caching.
+   *
+   * @param {object} log
+   * @param {object} config
+   */
+  constructor(log, config) {
+    this.log = log;
+    this.cacheTtlSeconds = config.subhub.plansCacheTtlSeconds;
```

Comment:
The default TTL is 10 minutes and I don't see an override in cloudops-deployment. Will folks be annoyed when they add a plan in stripe and they don't show up for up to 10 minutes, or is there a way to invalidate the cache via notifications possibly?

Reply:
Before we'll actually use this in prod, the new architecture will need to be approved. We will get Stripe webhook events directly, so we can hook into product/plan change events to refresh when Stripe is updated.